### ISSUE TITLE
ForeignKey non-nullable raises exception while creating

### ISSUE THREAD

**Issue by seanfarley** [encode/django-rest-framework#7344]
This is an issue based on #7258; quoted is the original question that has gone unanswered (the provided reply doesn't cover the non-nullable case)

> This use case is already covered by passing extra arguments to the serialiser's save

It's not covered by that at all! 
If the foreignkey isn't nullable then `create` will raise a ValidationError and `perform_create` will never be called.

Look at the definition of `create`:

```py
    def create(self, request, *args, **kwargs):
        serializer = self.get_serializer(data=request.data)
        serializer.is_valid(raise_exception=True)  # this fails before we can set extra args
        self.perform_create(serializer)
        headers = self.get_success_headers(serializer.data)
        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
```

_Originally posted by @JEphron in https://github.com/encode/django-rest-framework/pull/7258#issuecomment-610058099_

**Comment by rpkilby** [comment]
Hi @seanfarley. As described in the original issue, the FK value is inherently validated by the URL structure/view code. e.g., `/api/companies/<non-existent-id>/products`. If the ID doesn't exist, then the view should have already raised a 404, preempting the serializer validation. 

As to the serializer, since it is not responsible for validating the FK, its corresponding related field should be removed. If you still want the field present on serialized output, then the field needs to be marked as read-only instead. e.g.,

```python
class ProductSerializer(ModelSerializer):
    class Meta:
        model =  Product
        fields = ['company', ...]
        read_only_fields = ['company']
```

Then, you can supply the FK on creation.

Note that I'm closing, because I don't think there's a change we'd want to make here, but am happy to continue discussing the issue.

**Comment by seanfarley** [comment]
Thanks for the explanation and I understand closing; I do appreciate the nice response :-)

Following your code, I get the same error as that PR brought up; namely,

```
django.db.utils.IntegrityError: NOT NULL constraint failed: env.company_id
```

My model is nothing special:

```
class Env(models.Model):
    name = models.CharField(max_length=255)
    company = models.ForeignKey(Company, on_delete=models.CASCADE)
```

And the serializer is also very boring; it's just as you pasted. To be clear, I'm not sending the company id from the curl command because it's already in the URL `/api/companies/3/env`:

```
curl -X POST http://localhost:8000/api/companies/3/env -F name=foo
```

**Comment by rpkilby** [comment]
At this point, you're not hitting a validation error - you're hitting a DB error, which means that you haven't provided the FK value to the serializer on save. You'll need to overwrite the `perform_create` method on your view. e.g., 

```python
    def perform_create(self, serializer):
        company = self.get_company()
        serializer.save(company=company)  # ProductSerializer 
```

In the above, `get_company()` is just a stand in for however you would get the related object.

**Comment by seanfarley** [comment]
Oh geez; with all my code changing I got lost and forgot about overloading that method. I appreciate the thoughtful response, thank you very much!
